### ISSUE TITLE
Strange problems with 'less' program on Linux/macOS

### ISSUE THREAD

**Issue by amejia1** [sourcelair/xterm.js#977]
I have an Atom plugin named 'atom-xterm' which provides terminal emulators in the editor. On Linux and macOS, the less program acts strange. Upon initially starting up less (for example `less --help`), part of the top portion of the help doc that is displayed is cut off. Also the bottom portion leaves a bunch of blank lines. Also, when you move up or down, you can see portions in the middle of the doc get cut off. Upon quiting and running less again however, less works fine. This happens whenever a new terminal is started or a terminal is resized. Below is a gif on my Ubuntu VM showing exactly what I mean.

![xterm-less-problem](https://user-images.githubusercontent.com/86134/30306688-a932b6dc-9747-11e7-8eac-43f90ef85359.gif)

This happens consistently on any Linux (at least Ubuntu) or macOS machine. Other terminal emulator program of course don't have this problem. Also, the 'less' program available through Git for Windows works fine.

## Details
- Browser and browser version: This is actually happening in the Atom editor. Here are the details `atom --version`.
Atom    : 1.19.7
Electron: 1.6.9
Chrome  : 56.0.2924.87
Node    : 7.4.0

- OS version: Details from `lsb_release -a`
Distributor ID: Ubuntu
Description:    Ubuntu 17.04
Release:        17.04
Codename:       zesty

- xterm.js version:
v2.9.2

### Steps to reproduce

1. In the latest Atom editor with the latest atom-xterm plugin, on Linux or macOS, start `less --help`.
2. Start moving around the displayed doc to see the issue.
3. Close less, then reopen.
4. Find that less now almost displays correctly save for the top line which still shows the output from the current shell.

**Comment by Tyriar** [comment]
Looking at the gif, it looks like the pty is not sized correctly and is out of sync with xterm.js' size.

**Comment by amejia1** [comment]
I was just looking into this. This problem is not an issue with the v3 branch. I guess I'll just have to wait until v3 is released.

**Comment by amejia1** [comment]
Just to give you more info, this happens when a new terminal is opened and every time it's resized. Again though, no longer a problem with the v3 branch.

**Comment by Tyriar** [comment]
I guess we'll close it off then as we're working on v3 currently.

**Comment by gtritchie** [comment]
Ancient history at this point, but I needed to fix this for xterm 2.9.2 for RStudio; see this for how I did it in case someone else has a need for it: https://github.com/rstudio/xterm.js/commit/597a32d2fb3b5076fb21b2ef6173e3c1bb1d76bb

The bug was triggered as follows:
- initially created terminal 80x25
- immediately resized via fit.js (and a message to the pseudoterminal) to actual size
- if rows was less than 25, then running something using `less` such as `man man` then trying to page down (e.g. spacebar) would throw a Javascript exception and break rendering, and get terminal stuck in alt-buffer; other variations could happen including large blank areas in rendered page
- Reason: `this.scrollBottom` for alt-buffer was still set to reflect initial size of 25, it didn't get updated with the resize because alt buffer hadn't been created yet

This was already fixed in _essentially_ the same way in 3.x.

**Comment by Tyriar** [comment]
@gtritchie any reason you didn't update to 3?

**Comment by gtritchie** [comment]
Lack of time. I will do it in our next major release, but getting late in the cycle for the release that's in flight. We're a desktop product and a purely web product (desktop is a Qt frame using Chromium via QtWebEngine), and our JavaScript is mostly generated from Java via Gwt. Translation: things that should seem trivial sometimes aren't.

I did a brief experiment when 3.0 came out and hit two things (which I haven't revisited yet):

- `theming`; we did it via .css files generated by our build, so I need to do some work there to adapt to the new-improved model
- `require.js`: our code barfed on that; we have other dependencies, some quite old, which impose constraints there; I see there was some work on that in subsequent 3.x, so maybe that's already solved for me

Anyway, thanks for the great work!
